### ISSUE TITLE
Ordering of the B1+ images for B1map estimations on demo data

### ISSUE THREAD

**Issue by ChristophePhillips** [hMRI-group/hMRI-toolbox#23]
## INTRODUCTION

So we are currently updating the toolbox to make it compatible with the BIDS format as described [here](https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#anatomy-imaging-data) and [here](https://bids-specification.readthedocs.io/en/stable/99-appendices/11-qmri.html). One requirement is that the multi-FA images for the B1+ mapping are sorted in ascending order of the the FA value.

This constraint should not be affecting the maps estimation, as long as the associated metadata are linked to the correct images:
- **Good** ▶️   Overall it works nicely. 😁 
- **Not so good** ▶️  There exist small discrepancies between maps estimation depending on how the B1+ mapping (2 x 11) images are ordered!

## PROBLEM 

The quality assessment JSON file tells that for
- the original data (decreasing FA values)
   ````
   "SDR2s": {
		"MTw": 7.167988476039233,
		"PDw": 5.108226510523747,
		"T1w": 3.926471674618434
	},
	"PD": {
		"mean": 12047.88985555131,
		"SD": 723.2460970077426
	}
   ````
- the BIDS data (ascending)
  ````
  "SDR2s": {
		"MTw": 7.168852341561993,
		"PDw": 5.107599745898029,
		"T1w": 3.926375370419377
	},
	"PD": {
		"mean": 12038.24350601357,
		"SD": 772.7133611555393
	}
  ````
There are small difference in the SDR2s values at the 3rd decimal and the mean/std `PD` values are a bit off by <1%.

Looking at the images, still in the "Supplementary" folder, the only difference was in the `B1map` images. See this figure with the `B1map` (*top*) calculated from the original data (*top-left*) and BIDS formatted ones (*top-right*), and the `B1ref` (*bottom*) calculated from the original data (*bottom-left*) and BIDS formatted ones (*bottom-right*). Note the small discrepancy in the intensity and the different masking of the `B1map` image.
![comparison_B1map_B1ref](https://user-images.githubusercontent.com/2011934/114743390-f0535300-9d4c-11eb-8cd8-5f5506384a77.png)

### Main issue
This is not catastrophic but rather annoying as with the **same data and metadata**, one can end up with slightly different results!

## EXPLANATION?

1. When calculating the B1 map values `Y`, on [this line](https://github.com/hMRI-group/hMRI-toolbox/blob/master/hmri_create_b1map.m#L250), the `acos` of a ratio of values from the 2 echo images is used, then divided by the flip angle corresponding to the pair of images.

2. The problem stems from the masked out data, i.e. 0 in the lower frontal part of the images, the ration does not end up as `Nan` because a small `eps` is added to the numerator, therefore with the `acos(0)=pi/2` the value `Y` obtained is inversely proportionate to the flip angle, i.e. from 115 to 65 or *vice versa* depending on how the data are sorted (BIDS should use ascending values).

3. for each voxel, among the 11 pair of images, 5 pairs are selected picking up the largest value of the 1st echo image, i.e. after sorting the values, pick the last 5. However for the masked out voxels, the signal is zero so there is no ordering, so the selection will be those of the last 5 pairs, i.e. with the largest (resp. smallest) `Y` values for decreasing (resp. increasing) FA values!

4. For the voxel where the voxel value is 0, the ensuing mean and standard deviation (SD) estimated, on [this line](https://github.com/hMRI-group/hMRI-toolbox/blob/master/hmri_create_b1map.m#L273), will thus be larger (resp. smaller) for the for decreasing (resp. increasing) FA values. 

5. Finally at a later stage, the estimated maps are masked out, in [this function](https://github.com/hMRI-group/hMRI-toolbox/blob/master/hmri_create_B1Map_process.m), based on 2 criteria, 
   - Hertz field with a range [-110 110]
   - SD map lower than some threshold, set to 5

   Since with the increasing FA values (BIDS formatted), the SD is smaller and fewer voxels are masked out, and the resulting estimated B1 maps is different!

### Main question
How could we sort this out such that the results are the same irrespective of the ordering of the input data?

One option could be to ensure that the values are adjusted in [`hmri_create_b1map.m`](https://github.com/hMRI-group/hMRI-toolbox/blob/master/hmri_create_b1map.m) or [`hmri_create_B1Map_process.m`](https://github.com/hMRI-group/hMRI-toolbox/blob/master/hmri_create_B1Map_process.m) to account for the masked out voxels (with value set to 0).

**Comment by ChristophePhillips** [comment]
## SOLUTION?
And some further rambling...

Effectively there are 2 issues:
- the discrepancy in the results, i.e. lack of reproducibility
- the parameter of the method, i.e. using 5 (or another number) of echoes

For the latter, I am no expert in that area but I assume it is a compromise between "using as many as possible" and "avoiding noisy data". So I am not really keen on changing that parameter itself, like selecting all echoes of sorting them differently. At least not without the approval of one of the developers and a good scientific reason to do so.

Regarding the discrepancy, this should simply NOT occur, whatever the parameter used (5 or any other number). In fact it does NOT make sense to end up with a value where there was actually NO signal to begin with: the obtained value simply comes from taking the `acos` of a ratio which should be `0/0 = NaN` but is regularized into `0/(0+eps) = 0`, i.e. we end up with `pi/2` instead of `NaN`, which are further mingled with the flip angle.

So IMO it would make more sense to explicitly mask out those areas without signal. Still this will alter a little bit back-reproducibility: with a slightly different mask, re-estimation of the maps would lead to slightly different results. Still I think it would be a safer way in the long term... Your thought?

**Comment by mfcallaghan** [comment]
Hi @ChristophePhillips - the issue seems to stem from the masking, why is this different depending on the ordering?

**Comment by ChristophePhillips** [comment]
Ha, yes, tough one and it took me a while to figure out what was happening!

Here is an dummy example in Matlab code for you to try:
````
% Set FA values
FA_orig = 115:-5:65;
FA_bids = fliplr(FA_orig);
Npt = numel(FA_orig);

% Set 2 voxels, one at 0 the other random, for echo1 & echo2
% e2 is up to 50% larger than e1, again at random, such that e1/e2<1
e1_vox_orig = [rand(1,Npt)*100 ; zeros(1,Npt)] 
e2_vox_orig = e1_vox_orig.*([1 1]'*(1+rand(1,Npt)*5)) 
e1_vox_bids = fliplr(e1_vox_orig);
e2_vox_bids = fliplr(e2_vox_orig);

% Create that vector Y used for B1 estimation
Y_orig = acos(e1_vox_orig./(e2_vox_orig+.0001))/pi*180./([1 1]'*FA_orig)
Y_bids = acos(e1_vox_bids./(e2_vox_bids+.0001))/pi*180./([1 1]'*FA_bids)

% Sort thigns according to e1 values, along the 11 FAs
% and ick the 5 FAs with largest signal
[e1_vox_orig_s,sind_e1_vox_orig] = sort(e1_vox_orig,2);
[e1_vox_bids_s,sind_e1_vox_bids] = sort(e1_vox_bids,2);

% List the top 5 FAs leadign to largest voxel values
l_5largest_orig = sind_e1_vox_orig(:,end-4:end)
l_5largest_bids = sind_e1_vox_bids(:,end-4:end)

% Get the SD of Y for those 5 FAs
SD_Y_orig = [ ...
    std(Y_orig(1,l_5largest_orig(1,:))) ;
    std(Y_orig(2,l_5largest_orig(2,:))) ]
SD_Y_bids = [ ...
    std(Y_bids(1,l_5largest_bids(1,:))) ;
    std(Y_bids(2,l_5largest_bids(2,:))) ]
````
the final result will be this
````
SD_Y_orig =
    0.1827
    0.1288

SD_Y_bids =
    0.1827
    0.0651
````
Where there is signal, top, then we get the same SD value whatever the ordering of the FAs and it's all fine. Now when the signal is zero, bottom, then the SD is different! And that's exactly what happens in the hMRI code.

**Comment by antoinelutti** [comment]
hi Chris

I've thought about the problem that you flagged up but we happen to have our regular 'pull request' meeting next Wed.  at 1pm (we do, right people?). Would it not be easier to discuss this then?


Best



Antoine Lutti, PhD
Tenure-track Assistant Professor & Head of MRI Physics
Laboratoire de Recherche en Neuroimagerie (LREN)
Département des Neurosciences Cliniques (DNC)
Centre Hospitalier Universitaire Vaudois (CHUV)
University of Lausanne
Lausanne, Switzerland
Tel. +41 (0)21 314 8448 (internal ext 48448)
https://sites.google.com/view/antoinelutti/home
________________________________
De : Christophe Phillips ***@***.***>
Envoyé : mercredi, 21 avril 2021 14:08:18
À : hMRI-group/hMRI-toolbox
Cc : Lutti Antoine; Assign
Objet : Re: [hMRI-group/hMRI-toolbox] Ordering of the B1+ images for B1map estimations on demo data (#23)


Ha, yes, tough one and it took me a while to figure out what was happening!

Here is an dummy example in Matlab code for you to try:

% Set FA values
FA_orig = 115:-5:65;
FA_bids = fliplr(FA_orig);
Npt = numel(FA_orig);

% Set 2 voxels, one at 0 the other random, for echo1 & echo2
% e2 is up to 50% larger than e1, again at random, such that e1/e2<1
e1_vox_orig = [rand(1,Npt)*100 ; zeros(1,Npt)]
e2_vox_orig = e1_vox_orig.*([1 1]'*(1+rand(1,Npt)*5))
e1_vox_bids = fliplr(e1_vox_orig);
e2_vox_bids = fliplr(e2_vox_orig);

% Create that vector Y used for B1 estimation
Y_orig = acos(e1_vox_orig./(e2_vox_orig+.0001))/pi*180./([1 1]'*FA_orig)
Y_bids = acos(e1_vox_bids./(e2_vox_bids+.0001))/pi*180./([1 1]'*FA_bids)

% Sort thigns according to e1 values, along the 11 FAs
% and ick the 5 FAs with largest signal
[e1_vox_orig_s,sind_e1_vox_orig] = sort(e1_vox_orig,2);
[e1_vox_bids_s,sind_e1_vox_bids] = sort(e1_vox_bids,2);

% List the top 5 FAs leadign to largest voxel values
l_5largest_orig = sind_e1_vox_orig(:,end-4:end)
l_5largest_bids = sind_e1_vox_bids(:,end-4:end)

% Get the SD of Y for those 5 FAs
SD_Y_orig = [ ...
    std(Y_orig(1,l_5largest_orig(1,:))) ;
    std(Y_orig(2,l_5largest_orig(2,:))) ]
SD_Y_bids = [ ...
    std(Y_bids(1,l_5largest_bids(1,:))) ;
    std(Y_bids(2,l_5largest_bids(2,:))) ]


the final result will be this

SD_Y_orig =
    0.1827
    0.1288

SD_Y_bids =
    0.1827
    0.0651


Where there is signal, top, then we get the same SD value whatever the ordering of the FAs and it's all fine. Now when the signal is zero, bottom, then the SD is different! And that's exactly what happens in the hMRI code.

**Comment by DerOrfa** [comment]
fixed in pull request #28
